core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -0,0 +1,211 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.tasks;
+
+import hudson.AbortException;
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Action;
+import hudson.model.BuildListener;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildWrapper;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * A generalization of {@link BuildWrapper} that, like {@link SimpleBuildStep}, may be called at various points within a build.
+ * <p>Such a build wrapper would typically be written according to make few assumptions about how it is being used.
+ * @since TODO
+ */
+@SuppressWarnings("rawtypes") // inherited
+public abstract class SimpleBuildWrapper extends BuildWrapper {
+
+    /**
+     * Called when a segment of a build is started that is to be enhanced with this wrapper.
+     * @param context a way of collecting modifications to the environment for nested steps
+     * @param build a build being run
+     * @param workspace a workspace of the build
+     * @param launcher a way to start commands
+     * @param listener a way to report progress
+     * @param initialEnvironment the environment variables set at the outset
+     * @throws IOException if something fails; {@link AbortException} for user errors
+     * @throws InterruptedException if setup is interrupted
+     */
+    public abstract void setUp(Context context, Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException;
+
+    /**
+     * Parameter passed to {@link #setUp} to allow an implementation to specify its behavior after the initial setup.
+     */
+    public static final class Context {
+        private Disposer disposer;
+        private final Map<String,String> env = new HashMap<String,String>();
+        public @Nonnull Map<String,String> getEnv() {
+            return env;
+        }
+        /**
+         * Specify an environment variable override (as in {@link EnvVars#override}) to apply to processes launched within the block.
+         * If unspecified, environment variables will be inherited unmodified.
+         */
+        public void env(String key, String value) {
+            if (env.put(key, value) != null) {
+                throw new IllegalStateException("just one binding for " + key);
```

Comment:
True, but at least you would notice that you had a bug that ought to be fixed. Can change it to check before the `put` if you care.